cmd/thanos/query.go

```diff
@@ -475,7 +475,7 @@ func runQuery(
 			unhealthyStoreTimeout,
 			endpointInfoTimeout,
 		)
-		proxy            = store.NewProxyStore(logger, reg, endpoints.GetStoreClients, component.Query, selectorLset, storeResponseTimeout, store.RetrievalStrategy(grpcProxyStrategy))
+		proxy            = store.NewProxyStore(logger, reg, endpoints.GetStoreClients, component.Query, selectorLset, storeResponseTimeout, store.RetrievalStrategy(grpcProxyStrategy), queryReplicaLabels)
```

Comment:
Unfortunately, replica labels can also come through the API: https://github.com/thanos-io/thanos/blob/main/pkg/api/query/v1.go#L69. So,  it needs to go through the queryable to the Series() call on ProxyStore. `querier` contains:

```
...
type querier struct {
	ctx                 context.Context
	logger              log.Logger
	cancel              func()
	mint, maxt          int64
	replicaLabels       map[string]struct{}
	storeDebugMatchers  [][]*labels.Matcher
	proxy               storepb.StoreServer
```

Maybe we could change the type of `proxy` to `*store.ProxyStore` and then add another parameter to Series() call?

Reply:
Good catch, added replicas as part of the series message. I just wonder if there's a good way to ensure backwards compatibility, or if we should add a changelog entry advising people to first update stores.

One way I can think of is to add a boolean field to the info message indicating whether series are sorted or not. By default it will be false, so the querier can do its global sort if any single store has unsorted series.